### ISSUE TITLE
Web Resource should be able to  consume the body as a stream

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10190]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)| @adiroiban reported|
|-|-|
|Trac ID|trac#10190|
|Type|enhancement|
|Created|2021-05-08 10:09:07Z|

This is for the server-side.

Right now the Resource.render is called only after the whole body was received.

This make it complicated to implemented stream processing of the request.

And we have issues like [#4688](https://github.com/twisted/twisted/issues/4688) and [#4519](https://github.com/twisted/twisted/issues/4519)

My proposal is to have an updated API in which once all the headers are received, the request will call a new `Resource.headersReceived(request)` method.

The resource can then process the headers and for example inject a new `request.content` object that will support streaming ([#4688](https://github.com/twisted/twisted/issues/4688))

Or the resource can set a `request.content` that can handle a form request in a memory efficient way for ([#4519](https://github.com/twisted/twisted/issues/4519))

The resource can also reject the request to support 100-continue ([#6928](https://github.com/twisted/twisted/issues/6928))

---------

The main backward incompatible issues:

* Site.getResourceFor will be called before the body is processed. It might break traversal for code that are dispatching based on body content.

Solutions:

* Add a new `IStreamedResource` for which `getChildWithDefault` will be called after all the headers are received (but before the body is received). It will have a new `IStreamedResource.headersReceived(request)` method.
* Add a new `IRequest.setBodyConsumer` method that can be called to replace the current body consumeer.

Implementation issuse:

* `IStreamedResource.headersReceived` and `IStreamdedResource.getChildWithDefault` might need to return a deferred. For example when the request needs to be authenticated against an external service. 

------

It will need an explicit update for all existing resources but I think that this is fine

<details><summary>Searchable metadata</summary>

```
trac-id__10190 10190
type__enhancement enhancement
reporter__adiroiban adiroiban
priority__normal normal
milestone__None None
branch__ 
branch_author__ 
status__new new
resolution__None None
component__web web
keywords__None None
time__1620468547893934 1620468547893934
changetime__1620468547893934 1620468547893934
version__None None
owner__None None

```
</details>

**Comment by glyph** [comment]
Is this not a duplicate of #4688 ?

**Comment by adiroiban** [comment]
Thanks. Yes. It looks like a duplicate. #4688
